Create "Lifetime management and tread safety guarantees of Azure SDK clients" post (#1573)
Azure/azure-sdk: _posts/client-lifetime-management.md

# Lifetime management and tread-safety guarantees of Azure SDK clients

When using Azure SDK client in high throughput applications it's important to know how to maximize the performance while keeping the code correct and bug-free. 

The main rule of Azure SDK client lifetime management is - treat clients as singletons.

There is no need to keep more that than one instance of a client for a given set of constructor parameters. This can be implemented in many ways: creating an instance once and passing it around as a parameter, storing instance in a field, or registering it as a singleton in a dependency injection container of your choice.

Bad:
``` C#
foreach (var secretName in secretNames)
{
    var client = new SecretClient(new Uri("<secrets_endpoint>"), new DefaultAzureCredential());

>     KeyVaultSecret secret = client.GetSecret(secretName);

Author: What do you mean by burned?

>     KeyVaultSecret secret = client.GetSecret(secretName);

Reviewer: > We have an implicit cast from `Response<T>` to `T` that has been problematic in past studies.  We always talk about developers copy/pasting code samples, but a lot of them retype the sample by hand.  It makes intuitive sense because the sample is scoped to a different resource instance and you can change it bit by bit as you type.  This is a problem when we’re using the implicit cast in our samples (which we’ve been doing thus far for simpler looking samples).  An example like `SearchResults<Hotel> response = await client.SearchAsync(…);` that someone types as `var response = await client.SearchAsync(…);` will bomb on the next line when they type `response.GetResultsAsync()` instead of `response.Value.GetResultsAsync()`.  We saw this during the dry run and I changed the sample documentation to use the full `Response<T>` everywhere and nobody else had this problem during the study.  The implicit cast was mostly used accidentally during the study – folks would try to return a `var response` of type `Response<long>` from a `long` method and the cast made it just work even though the user wasn’t aware of it.

>     KeyVaultSecret secret = client.GetSecret(secretName);

Reviewer: That thing where people try to retype it by hand on their own but just use `var` because they don't see how it's different.

>     KeyVaultSecret secret = client.GetSecret(secretName);

Author: I'm on the fence because using the `Response` makes this particular example look ugly because of `response.Value.Value`.

>     KeyVaultSecret secret = client.GetSecret(secretName);

Reviewer: I don't feel strongly either way.  This isn't a README sample where people are going to be copying things to get started so it's fine to leave it as-is.

    Console.WriteLine(secret.Value);
}
```

Good:

``` C#
var client = new SecretClient(new Uri("<secrets_endpoint>"), new DefaultAzureCredential());
foreach (var secretName in secretNames)
{
    KeyVaultSecret secret = client.GetSecret(secretName);
    Console.WriteLine(secret.Value);
}
```

Also good:

``` C#
public class SecretConfigurationSource
{
    private readonly SecretClient _client;

    public SecretConfigurationSource()
    {
        _client = new SecretClient(new Uri("<secrets_endpoint>"), new DefaultAzureCredential());
    }

    public string GetConfigurationValue(string settingName)
    {
        KeyVaultSecret secret = _client.GetSecret(settingName);
        return secret.Value;
    }
}

> ```

Author: This wasn't exactly about wrapping, more about creating an instance once and storing into a field when implementing application specific features.

> ```

Reviewer: Would this example work better as a static class that provides a `SecretClient` singleton that's lazily initialized and then another bit of code elsewhere showing how you'd call it?

}

```

**NOTE:** the input and output models of the client methods are non-thread-safe and can only be accessed by one thread a time.

The following sample illustrates a bug where accessing a model from multiple threads might cause an undefined behavior.

``` C#
KeyVaultSecret newSecret = client.SetSecret("secret", "value");

foreach (var tag in tags)
{
    // Don't use model type from parallel threads
    Task.Run(() => newSecret.Properties.Tags[tag] = CalculateTagValue(tag));
}

client.UpdateSecretProperties(newSecret.Properties);
```

# Immutable after creation

The other important quality that allows reusing client instances safely is the immutability after creation. This means that after the client is constructed the endpoint it connects to, the credential or other values passed via the client options can not be changed.

``` C#
var secretClientOptions = new SecretClientOptions()
{
    Retry = 
    {
        Delay = TimeSpan.FromSeconds(5)
    }
};

var mySecretClient = new SecretClient(new Uri("<...>"), new DefaultAzureCredential(), secretClientOptions);

// This has no effect on the mySecretClient instance
secretClientOptions.Retry.Delay = TimeSpan.FromSeconds(100);
```

> **NOTE:** important exception from this rule are some `TokenCredential` implementations that allow rolling the key after the client was created (https://docs.microsoft.com/en-us/dotnet/api/azure.azurekeycredential.update?view=azure-dotnet#Azure_AzureKeyCredential_Update_System_String_)

Author: Saying TokenCredential was wrong here, I'll change to credential types.

> **NOTE:** important exception from this rule are some `TokenCredential` implementations that allow rolling the key after the client was created (https://docs.microsoft.com/en-us/dotnet/api/azure.azurekeycredential.update?view=azure-dotnet#Azure_AzureKeyCredential_Update_System_String_)

Author: Would we link to guidelines that apply to other parts of this post?

> **NOTE:** important exception from this rule are some `TokenCredential` implementations that allow rolling the key after the client was created (https://docs.microsoft.com/en-us/dotnet/api/azure.azurekeycredential.update?view=azure-dotnet#Azure_AzureKeyCredential_Update_System_String_)

Reviewer: https://azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-service-methods-thread-safety would make a lot of sense to link somewhere so it's clear that we're enforcing this with our guidelines.

> **NOTE:** important exception from this rule are some `TokenCredential` implementations that allow rolling the key after the client was created (https://docs.microsoft.com/en-us/dotnet/api/azure.azurekeycredential.update?view=azure-dotnet#Azure_AzureKeyCredential_Update_System_String_)

Author: Added guideline links, including to https://azure.github.io/azure-sdk/general_implementation.html#general-config-behaviour-changes

# Why client are not disposable

One frequent question that comes up is why aren't HTTP-based Azure clients implementing `IDisposable` while internally using the HttpClient that is disposable. An answer to this question is that all Azure SDK clients, by default, use a single shared HttpClient instance and don't create any other resources that need to be actively freed.

``` C#
// Both clients reuse the shared HttpClient and don't need to be disposed
var blobClient = new BlobClient(new Uri(sasUri));
var blobClient2 = new BlobClient(new Uri(sasUri2));
```

> If you provide a custom instance of HttpClient to Azure clients you become responsible for managing the `HttpClient` and dispose it at the right time.

Reviewer: ```suggestion
If you provide a custom instance of `HttpClient` to an Azure client, you become responsible for managing the `HttpClient` lifetime and disposing it at the right time.  We recommend [following `HttpClient` best practices](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=netcore-3.1#remarks) when customizing the transport.
```

``` C#
var httpClient = new HttpClient();

var clientOptions = new BlobClientOptions()
{
    Transport = new HttpClientTransport(httpClient)
}

// Both client would use the HttpClient instance provided in clientOptions
var blobClient = new BlobClient(new Uri(sasUri), clientOptions);
var blobClient2 = new BlobClient(new Uri(sasUri2), clientOptions);

//...

// some time later
httpClient.Dispose();
```

# Using ASP.NET Core

> If you are using Azure SDK clients in ASP.NET Core, client lifetime management can be simplified by using the Microsoft.Extensions.Azure package that provides a way to seamlessly integration Azure clients with ASP.NET Core dependency injection and configuration systems. See the [Best practices for using Azure SDK with ASP.NET Core](https://devblogs.microsoft.com/azure-sdk/best-practices-for-using-azure-sdk-with-asp-net-core/) blog post for details.

Reviewer: ```suggestion
If you are using Azure SDK clients in an ASP.NET Core application, client lifetime management can be simplified with the Microsoft.Extensions.Azure package that provides seamless integration of Azure clients with the ASP.NET Core dependency injection and configuration systems.  See the [Best practices for using Azure SDK with ASP.NET Core](https://devblogs.microsoft.com/azure-sdk/best-practices-for-using-azure-sdk-with-asp-net-core/) blog post for details.
```

Reviewer: Nit - link to the Extensions package README or Nuget?

# Lifetime management and thread-safety guarantees of Azure SDK clients

> When using Azure SDK client libraries in high throughput applications, it's important to know how to maximize performance while keeping the code correct and bug-free. 

Reviewer 2: Can you add something short to state the problem here, to help draw it for readers?

Maybe something as simple as ... "if you're writing a high-throughput application, you won't want to slow it down with extra memory allocations and processing steps.  You'll also want to avoid bugs that could be introduced by accessing data from multiple threads.  This article discusses how to use the Azure SDK client libraries to maximize performance by ..."

This helps give me context for why you're telling me this.

The main rule of Azure SDK client lifetime management is: **treat clients as singletons.**

> There is no need to keep more that than one instance of a client for a given set of constructor parameters or client options. This can be implemented in many ways: creating an instance once and passing it around as a parameter, storing instance in a field, or registering it as a singleton in a dependency injection container of your choice.

Reviewer 2: You have section headers for thread-safety, immutability, etc. ... It'd be great to have a section header here as well.

❌ Bad (extra allocations and initialization):
``` C#
foreach (var secretName in secretNames)
{
    var client = new SecretClient(new Uri("<secrets_endpoint>"), new DefaultAzureCredential());
    KeyVaultSecret secret = client.GetSecret(secretName);
    Console.WriteLine(secret.Value);
}
```

✔️ Good:

``` C#
var client = new SecretClient(new Uri("<secrets_endpoint>"), new DefaultAzureCredential());
foreach (var secretName in secretNames)
{
    KeyVaultSecret secret = client.GetSecret(secretName);
    Console.WriteLine(secret.Value);
}
```

✔️ Also good:

``` C#
public class Program
{
    internal static SecretClient Client;

    public static void Main()
    {
        Client = new SecretClient(new Uri("<secrets_endpoint>"), new DefaultAzureCredential());
    }
}

public class OtherClass
{
    public string DoWork()
    {
        KeyVaultSecret secret = Program.Client.GetSecret(settingName);
        Console.WriteLine(secret.Value);
    }
}
```

> # Thread-safe

Reviewer 2: It looks like the takeaway of this section is "clients are threadsafe, models are not" ... can you highlight that more - either by adding it a the top or having sub-headers to communicate there are two ideas?

❌ Bad:
``` C#
KeyVaultSecret newSecret = client.SetSecret("secret", "value");

foreach (var tag in tags)
{
    // Don't use model type from parallel threads
    Task.Run(() => newSecret.Properties.Tags[tag] = CalculateTagValue(tag));
}

client.UpdateSecretProperties(newSecret.Properties);
```

If you really need to acces the model from different threads use a syncronization primitive.

✔️ Good:

``` C#
KeyVaultSecret newSecret = client.SetSecret("secret", "value");

foreach (var tag in tags)
{
    Task.Run(() =>
    {
        lock (newSecret)
        {
            newSecret.Properties.Tags[tag] = CalculateTagValue(tag);
        }
    );
}

client.UpdateSecretProperties(newSecret.Properties);
```

# Immutable after creation

Clients are immutable after being created which also makes them safe to share and reuse safely ([guideline](https://azure.github.io/azure-sdk/general_implementation.html#general-config-behaviour-changes)).  This means that after the client is constructed, you cannot change the endpoint it connects to, the credential, and other values passed via the client options.

> ``` C#

Reviewer 2: You have a pattern of "Good/Bad" from other sections -- would it make sense to either add a "Bad" example here, or just label this one "Good"?  I feel like it'd be helpful to get the idea across to convey what I **can't** do if the client is immutable.

var secretClientOptions = new SecretClientOptions()
{
    Retry = 
    {
        Delay = TimeSpan.FromSeconds(5)
    }
};

var mySecretClient = new SecretClient(new Uri("<...>"), new DefaultAzureCredential(), secretClientOptions);

// This has no effect on the mySecretClient instance
secretClientOptions.Retry.Delay = TimeSpan.FromSeconds(100);
```

> **NOTE:** important exception from this rule are credential type implementations that are required to support rolling the key after the client was created ([guideline](azure.github.io/azure-sdk/dotnet_introduction.html#dotnet-auth-rolling-credentials)). Examples of such types include [AzureKeyCredential](https://docs.microsoft.com/en-us/dotnet/api/azure.azurekeycredential.update?view=azure-dotnet#Azure_AzureKeyCredential_Update_System_String_), [StorageSharedKeyCredential](https://docs.microsoft.com/en-us/dotnet/api/azure.storage.storagesharedkeycredential.setaccountkey?view=azure-dotnet#Azure_Storage_StorageSharedKeyCredential_SetAccountKey_System_String_).

Reviewer 2: Newcomers may not know what motivates the need for rolling credentials - can you add something short to describe why you'd want this?  (Or link out to more info)

# Why clients are not disposable

> One question that comes up often is why aren't HTTP-based Azure clients implementing `IDisposable` while internally using an `HttpClient` that is disposable?  All Azure SDK clients, by default, use a single shared `HttpClient` instance and don't create any other resources that need to be actively freed.

Reviewer 2: After the question, can you add "The answer /explanation is" .. to queue people that you're providing the answer and not giving more context?

Author: Haha, Ted removed my "the answer is" part :)

> One question that comes up often is why aren't HTTP-based Azure clients implementing `IDisposable` while internally using an `HttpClient` that is disposable?  All Azure SDK clients, by default, use a single shared `HttpClient` instance and don't create any other resources that need to be actively freed.

Reviewer 2: I'm also a little confused with how the answer is stated ...
It makes me wonder - ok... all of my clients are using the same shared `HttpClient` ... if I don't need to manage that, when does it go away?

``` C#
// Both clients reuse the shared HttpClient and don't need to be disposed
var blobClient = new BlobClient(new Uri(sasUri));
var blobClient2 = new BlobClient(new Uri(sasUri2));
```

> If you provide a custom instance of `HttpClient` to an Azure client, you become responsible for managing the `HttpClient` lifetime and disposing it at the right time.  We recommend [following `HttpClient` best practices](https://docs.microsoft.com/en-us/dotnet/api/system.net.http.httpclient?view=netcore-3.1#remarks) when customizing the transport.

Reviewer 2: This feels like it belongs in a sub-section ... I think the message here is "our clients aren't disposable ... if you provide one, you have to dispose it."  Again, maybe call this out in the section intro or with sub-headers?

``` C#
var httpClient = new HttpClient();

var clientOptions = new BlobClientOptions()
{
    Transport = new HttpClientTransport(httpClient)
}

// Both client would use the HttpClient instance provided in clientOptions
var blobClient = new BlobClient(new Uri(sasUri), clientOptions);
var blobClient2 = new BlobClient(new Uri(sasUri2), clientOptions);

//...

// some time later
httpClient.Dispose();
```

# Using ASP.NET Core

> If you are using Azure SDK clients in an ASP.NET Core application, client lifetime management can be simplified with the Microsoft.Extensions.Azure package that provides seamless integration of Azure clients with the ASP.NET Core dependency injection and configuration systems.  See the [Best practices for using Azure SDK with ASP.NET Core](https://devblogs.microsoft.com/azure-sdk/best-practices-for-using-azure-sdk-with-asp-net-core/) blog post or [Microsoft.Extensions.Azure package readme](https://github.com/Azure/azure-sdk-for-net/blob/master/sdk/core/Microsoft.Extensions.Azure/README.md) for details.

Reviewer 2: Love it.
